### DISCUSSION TITLE
Maybe use command line arguments to `abaqus cae` command

### DISCUSSION THREAD

**Discussion by diogo-rossi** [haiiliin/abqpy#1131] [Ideas]
This "final idea" came while concluding the #1104. 

The question I ask myself is: What if I want to use the `session.openOdb` method without a GUI?
(i.e. `abaqus cae -noGUI script.py` inside the `session.openOdb`)

Another question was: what if I want to use the `mdb.saveAs` method with a GUI? (`abaqus cae script=script.py` in the `mdb.saveAs`)

Then I thought that all the abaqus command line options could be set as arguments to run the script.

The main option `script=...` could be fixed, and the others inputted as arguments to the python script.

One possible approach would be:

In both `mdb.saveAs` and `session.openOdb`:
```python
os.system(f"{abaqus} cae script={sys.argv[0]} {sys.argv[1]} {sys.argv[2]} ... ")
```
Then, if I want a `noGUI` run, I would call:
```bash
python3 myscript.py -noGUI
```
But the other options would also be possible, including the removed `database` option:
```
abaqus cae
    [database=database-file]
    [replay=replay-file]
    [recover=journal-file]
    [startup=startup-file]
    [script=script-file]
    [noGUI=noGUI-file]
    [noenvstartup]
    [noSavedOptions]
    [noSavedGuiPrefs]
    [noStartupDialog]
    [custom=script-file]
    [guiTester=GUI-script]
    [guiRecord]
    [guiNoRecord]
```

**Comment by haiiliin** [comment]
Seems a great idea. But I have a question, how to implement the key-word arguments, i.e., if I want to use:
```sh
python myscript.py database=database-file
```
`sys.argv[i]` will be a `key-word` pair string, am I right?

**Reply by diogo-rossi** [reply]
Yes, that's right, the options would be pair strings.
And maybe we do a check: If there is a `=` in the `sys.argv[i]`, it would needed to be split (example `noGUI=file.py`)

**Reply by diogo-rossi** [reply]
Will need to implement the rules of the keywords [found here](https://help.3ds.com/2022/english/dssimulia_established/SIMACAEEXCRefMap/simaexc-c-caeproc.htm?contextscope=all&id=e1d3d62659e044aeb7d9a124426cf6a4). Maybe restrict to a number of specific keywords.

**Reply by haiiliin** [reply]
Why we need to split, can't we just use something like `abaqus cae script=filePath sys.argv[1]`?

**Reply by haiiliin** [reply]
I am thinking we can pass all the arguments to `abaqus` except for `script=...` and `sys.argv[0]`, let the `abaqus` command decide whether it is correct

**Reply by diogo-rossi** [reply]
Yes, you right.
I was thinking about the situation of `abaqus cae -noGUI filePath` (without `=`, with a space) but we may restrict the `sys.argv[1]` to be always `keyword=something`, that is simpler.

**Reply by haiiliin** [reply]
Why it should be always `key-word` option?

**Reply by diogo-rossi** [reply]
> Why it should be always `key-word` option?

Not always, that was just an example of keyword option that came splitted.
I think the only fixed keyword would be `script`

**Reply by haiiliin** [reply]
So we could do something like this, right?
```python
import os
import sys

args = ' '.join([arg for arg in sys.argv[1:] if not arg.split('=')[0] == 'script'])
os.system(f'abaqus cae script=script.py {args}')
```

**Reply by haiiliin** [reply]
One thing is, that we may want to use the `-noGUI` option by default.

**Comment by haiiliin** [comment]
Another possible solution is, rather than using the Python API to call the abaqus command, we can also create an Abaqus/Python Runner extension for VSCode to directly run the abaqus command.

**Reply by diogo-rossi** [reply]
That would be very nice!

But I think it is also good to maintain the call of Abaqus command from the library, because it is not dependent of VS Code.

Actually, VS Code has some issues in rendering the doc strings with Sphinx sintaxe (`.. code-block::`, `.. note::`, etc), that I think may be better in PyCharm.

### An generic snippet
![image](https://user-images.githubusercontent.com/57204023/189507216-41c60b41-1e97-49de-8dc6-7d8765f0681c.png)

### VS Code's render
![image](https://user-images.githubusercontent.com/57204023/189507238-0e8553a1-9eab-40b5-a47b-6e2a0d0a4b3a.png)

**Reply by haiiliin** [reply]
PyCharm can recongize `.. note::` but not `.. code-block::`

<img width="527" alt="image" src="https://user-images.githubusercontent.com/47051427/189507897-9f718ab7-cd6f-47a4-95e9-b2ddbd458234.png">

**Reply by haiiliin** [reply]
For directives that not recognized by PyCharm, it just ignore the directives and render the content.

**Reply by diogo-rossi** [reply]
Yeah, VS Code doesn't even show the content of these directives . . .

**Reply by diogo-rossi** [reply]
I just found out that VS Code render the content of code blocks using others ways to [show code example](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#showing-code-examples) with Sphinx.



The [reST literal blocks](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#rst-literal-blocks) and the [reST doctest blocks](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#doctest-blocks) works fine.

![image](https://user-images.githubusercontent.com/57204023/189513687-0cdb841a-cbaa-464b-b6ee-9a075e41ed07.png)

**Reply by haiiliin** [reply]
Wow, that is simple and fantastic!

**Reply by haiiliin** [reply]
> I just found out that VS Code render the content of code blocks using others ways to [show code example](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#showing-code-examples) with Sphinx.
> 
> The [reST literal blocks](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#rst-literal-blocks) and the [reST doctest blocks](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#doctest-blocks) works fine.
> 
> ![image](https://user-images.githubusercontent.com/57204023/189513687-0cdb841a-cbaa-464b-b6ee-9a075e41ed07.png)

Create an issue in #1265.

**Comment by haiiliin** [comment]
Instead of passing arguments to `python`, how about add a individual command line application `abqpy` and use:
```sh
abqpy script.py [options]
```
which I think is more natural.

**Reply by diogo-rossi** [reply]
You're right, it is more natural.
But I think we should provide argument passing to abaqus from python. Look at [these examples](https://help.3ds.com/2022/English/DSSIMULIA_Established/SIMACAEEXCRefMap/simaexc-c-caeproc.htm?contextscope=all&id=30dbc96d9b6c4ac9a12a9b72015b6ee2#simaexc-c-caeproc-t-abqexecprocexample1).

Imagine I have a file `scriptArgs.py` with the following code:

```python
# file `scriptArgs.py`
import sys
from abaqus import *

if sys.version_info[0] < 3: # run this only in abaqus (python 2.7)
    with open('executed_from_python2.txt','w') as fd:
        fd.write(sys.argv[-1])

mdb.saveAs('temp.cae')
```

If I run

```cmd
python3 scriptArgs.py argument1
```

it will pass the argument to the script inside abaqus.
In this case, it will create a output file `executed_from_python2.txt` with the information:

```
argument1
```

but could be use to pass arguments to abaqus python script.

**Reply by haiiliin** [reply]
So preferably, these several methods should be all implemented.

**Reply by diogo-rossi** [reply]
I've made some tests. Here is some conclusions:
Initially, we have 2 options to pass arguments to Abaqus through python:

1.  Pass all the arguments to Abaqus (except for `script=...` and `sys.argv[0]`), and let the Abaqus command decide whether it is correct:
    A possible implementation would be, as you pointed:

    ```python
    import os, sys
    args = ' '.join([arg for arg in sys.argv[1:] if not arg.split('=')[0] == 'script'])
    os.system(f'abaqus cae script={sys.argv[0]} {args}')
    ```
    And the call of the script (ex: `script.py`) would be:
    ```cmd
    python3 scripty.py arg1 arg2 arg3 ...
    ```
    **Example**: If I wanted to run with `noGUI` option and pass `argument1` and `argument2` to the script, I would need to call:
    ```cmd
    python3 scripty.py -noGUI -- argument1 argument2
    ```
2. Add a individual command line application `abqpy` instead of passing arguments to `python`, and use it to pass arguments.
    We will also have to implement the Abaqus command to pass the arguments, but it will be called naturally using:
    ```cmd
    abqpy [option1 option2 ...] script.py [arg1 arg2 arg3]
    ```
    And, for example, if I wanted to run with `noGUI` option and pass `argument1` and `argument2` to the script, I would need to call:
    ```cmd
    abqpy -noGUI script.py argument1 argument2
    ```

**Reply by diogo-rossi** [reply]
The 2 solutions have pros and cons.

The advantage of **Option 1** is that you could configure the IDE (VS Code, PyCharm) to launch the python script including any option and any arguments (mostly IDE have a "launch button").

The advantage of **Option 2** is that the command line separates what is an Abaqus CAE "option" and what is the script "argument" without the use of the option `--`, which is more natural.

I'm thinking solution 1 is better. But, in either solution, we will basically write an application that replaces the call of Abaqus CAE:
```cmd
abaqus cae [script.py option1 option2 ...] -- [arg1 arg2 arg3]
```
Which brings no real gain after all.

**Reply by diogo-rossi** [reply]
So, I'm thinking of a **Option 3**, that may be applicable to the first one to reduce the call in `python`. But, uses a global variable.

It is similar to one I saw in another library. The proposal is:

In the main `__init__.py` module, we add a global dictionary for the "run time configuration parameters" with default values:

```python
rcParam = {'noGUI': True, 'database': None, 'replay': None, 'recover': None ...}
```

These options would then be used in the call of the Abaqus command, with the properly implementation.

So, inside the script, I could change any option I need before the methods `mdb.saveAs` or `session.openOdb`.

But, I would need to also import `abaqus` for this. Example:

```python
import abaqus
from abaqus import *

abaqus.rcParam = {'noGUI': False, 'database': <database-file> ... } # Change any default

...

mdb.saveAs('name.cae')
```

Finally, if I wanted to run again the same example with `noGUI` option and pass `argument1` and `argument2` to the script, I would call the simple line below, with only the arguments (without any options and without the `--`):
```cmd
python3 scripty.py argument1 argument2
```

What do you think of this proposal?

The great disadvantage of this solution is the need to `import abaqus` and change the global dictionary inside the script.

But that will not be necessary if we use the defaults, only to add some advanced option.

**Reply by haiiliin** [reply]
I think this method is applicable for those "config" options, to config how the package behaves during runtime, like the `noGUI` option. But for those "dynamic" options like `database`, it shouldn't be processed like this way, for example, `matplotlib` will never  store the `x-y` data into the `rcParam` dictionary, right?

**Reply by diogo-rossi** [reply]
>But for those "dynamic" options like `database`, it shouldn't be processed like this way, for example, `matplotlib` will never store the `x-y` data into the `rcParam` dictionary, right?

I agree about the `matplotlib` data, but I'm not understanding why `database` is a "dynamic" option, different from `noGUI`.

I mean, the options are all "input once" (or "config" option), that are called from the `abaqus` command, rigth?

Example:
```
abaqus cae noGUI=script.py database=file.odb ... etc
```

is that right?

**Reply by diogo-rossi** [reply]
oh, just to clarify, this `database` config option is that one used in `abaqus cae` command line:
```
abaqus cae
    [database=database-file]
    [replay=replay-file]
    [recover=journal-file]
    [startup=startup-file]
    [script=script-file]
    [noGUI=noGUI-file]
    [noenvstartup]
    [noSavedOptions]
    [noSavedGuiPrefs]
    [noStartupDialog]
    [custom=script-file]
    [guiTester=GUI-script]
    [guiRecord]
    [guiNoRecord]
```

The argument used in the `session.openOdb(...)` is in deed dynamic and may change inside the script.

**Reply by haiiliin** [reply]
I mean the `database` can be any `.odb` file that may change over time, but for options like `noGUI` we'd probably use the same value always. For those options, we can actually use the `environment variable` like the `ABAQUS_BAT_PATH`.

**Reply by diogo-rossi** [reply]
> For those options, we can actually use the environment variable like the ABAQUS_BAT_PATH.

Yes, I thought of an environment variable too. But I couldn't think of a way to store all those options in one single environment variable.

>I mean the database can be any .odb file that may change over time, but for options like noGUI we'd probably use the same value always. 

Yes, `database` may be any `.odb` file, but actually, I think this option may not be used so many times.
This argument is used only to open Abaqus CAE with a pre-chosen ODB file (remember that we removed it in https://github.com/haiiliin/abqpy/commit/198b6d8db4bbae0f97531c77dc54e458c62e2f02) but we may choose an ODB any time inside the script, so think this `database` argument may not be used so much.

On the other hand, the `noGUI` option appeared when I made the questions in the beginning of the discussion:
> What if I want to use the `session.openOdb` method without a GUI? i.e. `abaqus cae noGUI=script.py`
> What if I want to use the `mdb.saveAs` method with a GUI? i.e. `abaqus cae script=script.py`

So I think that this argument may actually be more used.

**Reply by haiiliin** [reply]
> Yes, I thought of an environment variable too. But I couldn't think of a way to store all those options in one single environment variable.

We can define a environment variable like `ABAQUS_COMMAND_OPTIONS`, users can put all the options here, or define multiple individual environment variables like `ABAQUS_NOGUI`. 

Another option is to define `.env` file locally and use `dotenv` to load the local environment file in abaqus/__init__.py file

**Reply by haiiliin** [reply]
Options in https://github.com/haiiliin/abqpy/discussions/1131#discussioncomment-3621880 can be divided into two categories:

- Configrations:
  - `noenvstartup`
  - `noSavedOptions`
  - `noSavedGuiPrefs`
  - `noStartupDialog`
  - `guiRecord`
  - `guiNoRecord`
- Script-file-related options
  - `database`
  - `relay`
  - `recover`
  - `startup`
  - `noGUI`
  - `custom`
  - `guiTester`

The `noGUI=noGUI-file` seems to be equivalent to `-noGUI script=noGUI-file`.

**Reply by diogo-rossi** [reply]
> The `noGUI=noGUI-file` seems to be equivalent to `-noGUI script=noGUI-file`.

Yes. I think the two options below are mutually exclusive:
```python
# runs with GUI
abaqus cae script=file.py
```
```python
# runs without GUI
abaqus cae noGUI=file.py
```
That should be considered in implementation.

> We can define a environment variable like `ABAQUS_COMMAND_OPTIONS`, users can put all the options here, or define multiple individual environment variables like `ABAQUS_NOGUI`.
> 
> Another option is to define `.env` file locally and use `dotenv` to load the local environment file in abaqus/**init**.py file

Yes, those would be other alternatives, lets say, **Option 4** and **Option 5**.

Is `dotenv`  [this package](https://pypi.org/project/python-dotenv/)? It seems a very useful package.

**Reply by diogo-rossi** [reply]
Whichever will be the chosen option, I think the default configuration will need to include `noGUI` argument in [this line](https://github.com/haiiliin/abqpy/blob/76c8b20ed24ec6722d7ffbd1431814cfc01b289b/src/abaqus/__init__.py#L87), is that right?

**Reply by diogo-rossi** [reply]
The disadvantage of using environment variables is the need to change it in the system, outside the script, every time when new options are intended.

The disadvantage of using `.env` file is the same (external config) and the need of an external library, if it is really the case.

So, I think there is no clear better solution. But I think that the argument passing (values after `--`) should really be implemented.

One thing that may be clear in the documentation is that the `abaqus cae` command already passes predefined arguments to the script, so, for instance, the `sys.argv[1]` value passed to the `python3` interpreter will not be the same as the passed to the `abaqus python2` interpreter.

Therefore, the access of the arguments passed to the script with the line `python3 scripty.py arg1 arg2 arg3 ...` should be made inside the script in the same manner as the [examples in Abaqus documentation](https://help.3ds.com/2022/English/DSSIMULIA_Established/SIMACAEEXCRefMap/simaexc-c-caeproc.htm?contextscope=all&id=30dbc96d9b6c4ac9a12a9b72015b6ee2#simaexc-c-caeproc-t-abqexecprocexample1), i.e., `sys.argv[-3]`, `sys.argv[-2]`, `sys.argv[-1]`, `...`. That have to be clear in the documentation.

**Reply by haiiliin** [reply]
> Is dotenv this package? It seems a very useful package.

Yes

**Reply by haiiliin** [reply]
> Whichever will be the chosen option, I think the default configuration will need to include `noGUI` argument in [this line](https://github.com/haiiliin/abqpy/blob/76c8b20ed24ec6722d7ffbd1431814cfc01b289b/src/abaqus/__init__.py#L87), is that right?

Right, by default it should be executed in noGUI mode

**Reply by diogo-rossi** [reply]
> Right, by default it should be executed in noGUI mode

Ok. I can implement that.
May I also include the script's argument passing? (arguments after `--`)

I'll leave the `abaqus cae` command option passing to be defined, not included for now.
I think we need to choose between Options 3, Option 4 and Option 5.

**Reply by haiiliin** [reply]
> May I also include the script's argument passing? (arguments after `--`)
> 
> I'll leave the `abaqus cae` command option passing to be defined, not included for now.
> I think we need to choose between Options 3, Option 4 and Option 5.

That could be nice.

**Reply by diogo-rossi** [reply]
ok

**Comment by haiiliin** [comment]
@Diogo-Rossi BTW, thanks for so many contributions recently, maintaining this large project is really difficult. If you are interested and willing to make more contributions, I can invite you as a collaborator, then you can have more access to do things freely.

**Reply by diogo-rossi** [reply]
@haiiliin  Thank you for this message and invitation! I'm very appreciate, and I'm happy to help.
But, as I don't have much experience in development workflow (specially automatic merge, back-port compatibility and configuration settings) I'm not very secure in have write permissions to the project, because I'm afraid to "broke" something. So, at least for a while, if you don't mind, I would like to continue to submit revision to validation before merge to the main project

**Reply by haiiliin** [reply]
Ok, never mind. Still appreciate for your kind contributions recently.

**Reply by haiiliin** [reply]
@Diogo-Rossi Actually, you are not able to break anything without my permission, the branch main, V20** are protected and you can't push to them directly (even I can't push directly), the only way is to create pull requests, and the pull requests cannot be merged without my review.

**Reply by diogo-rossi** [reply]
> @Diogo-Rossi Actually, you are not able to break anything without my permission, the branch main, V20** are protected and you can't push to them directly (even I can't push directly), the only way is to create pull requests, and the pull requests cannot be merged without my review.

Oh, if it is like that, I'll be glad to be a collaborator. Thank you very much.

**Reply by haiiliin** [reply]
Send you an invitation

**Comment by haiiliin** [comment]
I have another thought, how about creating an independent package `abqpy` to store utils that are not abaqus-related, such as `mdb.saveAs()` and `session.openOdb()`, and import them in the `abaqus` package, I started a discussion in #1232.

**Reply by diogo-rossi** [reply]
That's an interesting idea. I'm not sure I understood. My question is, how would be the usage of the package?

**Reply by haiiliin** [reply]
Done in #1241, you can check it. 

Defined a function to run the abaqus command in [src/abqpy/abaqus.py](https://github.com/haiiliin/abqpy/blob/main/src/abqpy/abaqus.py), decorators were moved to [src/abqpy/decorators.py](https://github.com/haiiliin/abqpy/blob/main/src/abqpy/decorators.py).

**Reply by diogo-rossi** [reply]
I saw it. Very nice!

Do you think we should put `exit_after=True` also in `mdb.saveAs`?

I mean, to prevent running Abaqus several times, if one save the model multiple times.

```python
# do stuff

mdb.saveAs('model.cae')

# do stuff

mdb.saveAs('model.cae')

# do stuff

mdb.saveAs('model.cae')

```

**Reply by haiiliin** [reply]
If not, what will be influenced?

**Reply by diogo-rossi** [reply]
I thought if we do not put `exit_after=True` also in `mdb.saveAs` each line will call Abaqus.

```python
# contents of `script.py`

# do stuff

mdb.saveAs('model.cae') # execute `abaqus cae noGUI=script.py`

# do stuff

mdb.saveAs('model.cae') # execute `abaqus cae noGUI=script.py`

# do stuff

mdb.saveAs('model.cae') # execute `abaqus cae noGUI=script.py`
```

If we put, I think that only the first call of Abaqus will be made.

Is that right?

**Reply by haiiliin** [reply]
Yes. I just know what you mean, if we don't exit the first time, the script will be run several times, fair point, I think we need to do this.

**Reply by haiiliin** [reply]
Another point is, we may run the abaqus command in other methods, like `mdb.save()`.

**Reply by diogo-rossi** [reply]
Good point! Maybe the most used method will be `mdb.save()`, instead of `mdb.saveAs()`

**Reply by haiiliin** [reply]
> Another point is, we may run the abaqus command in other methods, like `mdb.save()`.

Created an issue in #1266.

**Comment by haiiliin** [comment]
Summarize this discussion about command line arguments in https://github.com/haiiliin/abqpy/issues/1269#issuecomment-1245533470.
